File: src-tauri/src/pipeline/port.rs
use std::collections::HashMap;
use std::sync::mpsc::{channel, Receiver, Sender};

use miette::{IntoDiagnostic, Result};

use crate::error::MrDamianError;
use crate::pipeline::message::Message;
use crate::pipeline::packet::Packet;

#[derive(Debug)]
pub struct InputPort {
    pub base_sender: Sender<Packet>,
    pub receiver: Receiver<Packet>,
}

impl InputPort {
    pub fn new() -> InputPort {
        let (base_sender, receiver) = channel::<Packet>();
        Self {
            base_sender,
            receiver,
        }
    }

    pub fn accquire(&self, dest: &str) -> OutputPort {
        OutputPort {
            dest: dest.to_string(),
            sender: self.base_sender.clone(),
        }
    }

    pub fn receive(&self) -> Result<Packet> {
        self.receiver.recv().into_diagnostic()
    }
}

#[derive(Debug)]
pub struct OutputPort {
    pub dest: String,
    pub sender: Sender<Packet>,
}

impl OutputPort {
    pub fn send(&self, message: Message) -> Result<()> {
        let packet = Packet {
            port: self.dest.clone(),
            message,
        };
        self.sender.send(packet).into_diagnostic()
    }
}

#[derive(Debug, Default)]
pub struct OutputPorts {
    pub ports: HashMap<String, Vec<OutputPort>>,
}

impl OutputPorts {
    pub fn attach(&mut self, src: &str, port: OutputPort) {
        self.ports
            .entry(src.to_string())
            .or_insert(Vec::new())
            .push(port);
    }

    pub fn send(&self, packet: Packet) -> Result<()> {
        let port = self
            .ports
            .get(&packet.port)
            .ok_or_else(|| MrDamianError::PortNotFound(packet.port.clone()))?;
        for p in port {
            p.send(packet.message.clone())?;
        }
        Ok(())
    }
}

File: operation/pipeline/port.go
package pipeline

import "github.com/yuniruyuni/mrdamian/model"

type (
	Sender   chan<- *Packet
	Receiver <-chan *Packet
)

type InputPort struct {
	BaseSender Sender
	Receiver   Receiver
}

type InputPorts map[string]InputPort

const PortBufferSize = 32

func NewInputPort() InputPort {
	ch := make(chan *Packet, PortBufferSize)
	return InputPort{
		BaseSender: ch,
		Receiver:   ch,
	}
}

func (p *InputPort) Accquire(dest string) OutputPort {
	return OutputPort{
		Dest:   dest,
		Sender: p.BaseSender,
	}
}

func (p *InputPort) Receive() (*Packet, error) {
	return <-p.Receiver, nil
}

type OutputPort struct {
	Dest   string
	Sender Sender
}

func (p *OutputPort) Send(message Message) {
	packet := Packet{
		Port:    p.Dest,
		Message: message,
	}
	p.Sender <- &packet
}

type OutputPorts map[string][]OutputPort

func NewOutputPorts() OutputPorts {
	return OutputPorts{}
}

func (p OutputPorts) Attach(src string, port OutputPort) {
	p[src] = append(p[src], port)
}

func (p OutputPorts) Send(packet *Packet) error {
	port, ok := p[packet.Port]
	if !ok {
		return model.ErrPortNotFound
	}

	for _, p := range port {
		p.Send(packet.Message)
	}

	return nil
}
